[resolver] [mvn] Fragment bundles are not resolved since 4.3.1. <Bundle> was not associated with a resource in the resolution

The last version of bnd maven plugin that resolves the Fragment bundle correctly is 4.3.1. After update to 5+ version maven plugin fires the error:

`[ERROR] Failed to execute goal biz.aQute.bnd:bnd-export-maven-plugin:5.2.0:export (default) on project gateway-distribution-bnd-assembly: The capability for wire osgi.identity;filter:='(osgi.identity=logback-smart-filter)' -> [logback-smart-filter version=20.10.0] was not associated with a resource in the resolution -> [Help 1]`

--- a/biz.aQute.bnd.gradle/bnd.bnd
+++ b/biz.aQute.bnd.gradle/bnd.bnd
@@ -22,8 +22,8 @@ bnd_version: ${replace;${bnd_plugin};.*:(.*);$1}
 # Use groovydoc task generated doc for -javadoc.jar
 -sources: false
 -maven-release: pom;path=JAR,\
- sources;path=${src},\
- javadoc;path=${target}/docs/groovydoc
+ sources;path="${src}",\
+ javadoc;path="${target}/docs/groovydoc"
 
 # Stop Eclipse whining about the bnd files using specific Bundle-SymbolicNames
 -fixupmessages.bndtools: "Eclipse: The Bundle Symbolic * name"

--- a/biz.aQute.bndlib/src/aQute/bnd/classindex/ClassIndexerAnalyzer.java
+++ b/biz.aQute.bndlib/src/aQute/bnd/classindex/ClassIndexerAnalyzer.java
@@ -31,13 +31,13 @@ public class ClassIndexerAnalyzer implements AnalyzerPlugin {
 			if (classspace.isEmpty())
 				continue;
 
-			List<Integer> hashes = new ArrayList<>();
+			List<Long> hashes = new ArrayList<>();
 			for (Clazz c : classspace) {
 				TypeRef tr = c.getClassName();
 				if (tr.isArray() || tr.isNested())
 					continue;
 
-				int hash = tr.getShorterName()
+				long hash = tr.getShorterName()
 					.hashCode();
 				hashes.add(hash);
 			}

--- a/biz.aQute.bndlib/src/aQute/bnd/osgi/resource/CapReqBuilder.java
+++ b/biz.aQute.bndlib/src/aQute/bnd/osgi/resource/CapReqBuilder.java
@@ -100,7 +100,8 @@ public class CapReqBuilder {
 			}
 		}
 
-		if (name.equals(ResourceUtils.getVersionAttributeForNamespace(getNamespace()))) {
+		if ((value instanceof aQute.bnd.version.Version)
+			|| name.equals(ResourceUtils.getVersionAttributeForNamespace(getNamespace()))) {
 			value = toVersions(value);
 		}
 

--- a/biz.aQute.bndlib/src/aQute/bnd/osgi/resource/ResourceBuilder.java
+++ b/biz.aQute.bndlib/src/aQute/bnd/osgi/resource/ResourceBuilder.java
@@ -757,13 +757,13 @@ public class ResourceBuilder {
 			return;
 		}
 
-		List<Integer> hashes = resources.stream()
+		List<Long> hashes = resources.stream()
 			.map(NamedNode::name)
 			.filter(Descriptors::isBinaryClass)
 			.map(Descriptors::binaryToSimple)
 			.distinct()
 			.filter(simple -> !Verifier.isNumber(simple))
-			.map(ClassIndexerAnalyzer::hash)
+			.map(simple -> Long.valueOf(ClassIndexerAnalyzer.hash(simple)))
 			.collect(toList());
 		if (hashes.isEmpty()) {
 			return;

--- a/biz.aQute.bndlib/src/aQute/bnd/osgi/resource/ResourceImpl.java
+++ b/biz.aQute.bndlib/src/aQute/bnd/osgi/resource/ResourceImpl.java
@@ -12,6 +12,7 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 
 import org.osgi.framework.namespace.IdentityNamespace;
 import org.osgi.resource.Capability;
@@ -98,33 +99,21 @@ class ResourceImpl implements Resource, Comparable<Resource>, RepositoryContent 
 		if (this == other)
 			return true;
 
-		if (other == null || !(other instanceof Resource))
+		if (!(other instanceof Resource))
 			return false;
 
 		Map<URI, String> thisLocations = getLocations();
-		Map<URI, String> otherLocations;
-
-		if (other instanceof ResourceImpl) {
-			otherLocations = ((ResourceImpl) other).getLocations();
-		} else {
-			otherLocations = ResourceUtils.getLocations((Resource) other);
-		}
+		Map<URI, String> otherLocations = (other instanceof ResourceImpl) ? ((ResourceImpl) other).getLocations()
+			: ResourceUtils.getLocations((Resource) other);
 
 		Collection<URI> overlap = retain(thisLocations.keySet(), otherLocations.keySet());
 
-		for (URI uri : overlap) {
-			String thisSha = thisLocations.get(uri);
-			String otherSha = otherLocations.get(uri);
-			if (thisSha == otherSha)
-				return true;
-
-			if (thisSha != null && otherSha != null) {
-				if (thisSha.equals(otherSha))
-					return true;
-			}
-		}
-
-		return false;
+		return overlap.stream()
+			.anyMatch(uri -> {
+				String thisSha = thisLocations.get(uri);
+				String otherSha = otherLocations.get(uri);
+				return Objects.equals(thisSha, otherSha);
+			});
 	}
 
 	private Map<URI, String> getLocations() {

--- a/biz.aQute.repository/src/aQute/bnd/repository/fileset/FileSetRepository.java
+++ b/biz.aQute.repository/src/aQute/bnd/repository/fileset/FileSetRepository.java
@@ -24,8 +24,6 @@ import org.osgi.util.promise.PromiseFactory;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import aQute.bnd.osgi.Constants;
-import aQute.bnd.osgi.Domain;
 import aQute.bnd.osgi.Jar;
 import aQute.bnd.osgi.repository.BaseRepository;
 import aQute.bnd.osgi.repository.BridgeRepository;
@@ -42,7 +40,6 @@ import aQute.bnd.version.MavenVersion;
 import aQute.bnd.version.Version;
 import aQute.lib.exceptions.Exceptions;
 import aQute.lib.strings.Strings;
-import aQute.libg.cryptography.SHA256;
 import aQute.maven.api.Revision;
 import aQute.maven.provider.POM;
 import aQute.service.reporter.Reporter;
@@ -99,38 +96,38 @@ public class FileSetRepository extends BaseRepository implements Plugin, Reposit
 				return null;
 			}
 			ResourceBuilder rb = new ResourceBuilder();
-			try (Jar jar = new Jar(file)) {
-				Domain manifest = Domain.domain(jar.getManifest());
-				boolean hasIdentity = rb.addManifest(manifest);
+			try {
+				boolean hasIdentity = rb.addFile(file, null);
 				if (!hasIdentity) {
-					Optional<Revision> revision = jar.getPomXmlResources()
-						.findFirst()
-						.map(asFunctionOrElse(pomResource -> new POM(null, pomResource.openInputStream(), true), null))
-						.map(POM::getRevision);
+					try (Jar jar = new Jar(file)) {
+						Optional<Revision> revision = jar.getPomXmlResources()
+							.findFirst()
+							.map(asFunctionOrElse(pomResource -> new POM(null, pomResource.openInputStream(), true),
+								null))
+							.map(POM::getRevision);
 
-					String name = jar.getModuleName();
-					if (name == null) {
-						name = revision.map(r -> r.program.toString())
-							.orElse(null);
+						String name = jar.getModuleName();
 						if (name == null) {
-							return null;
+							name = revision.map(r -> r.program.toString())
+								.orElse(null);
+							if (name == null) {
+								return null;
+							}
 						}
-					}
 
-					Version version = revision.map(r -> r.version.getOSGiVersion())
-						.orElse(null);
-					if (version == null) {
-						version = new MavenVersion(jar.getModuleVersion()).getOSGiVersion();
-					}
+						Version version = revision.map(r -> r.version.getOSGiVersion())
+							.orElse(null);
+						if (version == null) {
+							version = new MavenVersion(jar.getModuleVersion()).getOSGiVersion();
+						}
 
-					CapReqBuilder identity = new CapReqBuilder(IdentityNamespace.IDENTITY_NAMESPACE)
-						.addAttribute(IdentityNamespace.IDENTITY_NAMESPACE, name)
-						.addAttribute(IdentityNamespace.CAPABILITY_VERSION_ATTRIBUTE, version)
-						.addAttribute(IdentityNamespace.CAPABILITY_TYPE_ATTRIBUTE, IdentityNamespace.TYPE_UNKNOWN);
-					rb.addCapability(identity);
+						CapReqBuilder identity = new CapReqBuilder(IdentityNamespace.IDENTITY_NAMESPACE)
+							.addAttribute(IdentityNamespace.IDENTITY_NAMESPACE, name)
+							.addAttribute(IdentityNamespace.CAPABILITY_VERSION_ATTRIBUTE, version)
+							.addAttribute(IdentityNamespace.CAPABILITY_TYPE_ATTRIBUTE, IdentityNamespace.TYPE_UNKNOWN);
+						rb.addCapability(identity);
+					}
 				}
-				rb.addContentCapability(file.toURI(), SHA256.digest(file)
-					.asHex(), file.length(), hasIdentity ? Constants.MIME_TYPE_BUNDLE : Constants.MIME_TYPE_JAR);
 			} catch (Exception f) {
 				return null;
 			}

--- a/biz.aQute.resolve/src/biz/aQute/resolve/AbstractResolveContext.java
+++ b/biz.aQute.resolve/src/biz/aQute/resolve/AbstractResolveContext.java
@@ -1,6 +1,8 @@
 package biz.aQute.resolve;
 
 import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toCollection;
+import static java.util.stream.Collectors.toList;
 import static org.osgi.framework.Constants.SYSTEM_BUNDLE_SYMBOLICNAME;
 import static org.osgi.framework.namespace.BundleNamespace.BUNDLE_NAMESPACE;
 import static org.osgi.framework.namespace.HostNamespace.HOST_NAMESPACE;
@@ -221,22 +223,20 @@ public abstract class AbstractResolveContext extends ResolveContext {
 			// repos.
 			boolean optional = Namespace.RESOLUTION_OPTIONAL.equals(requirement.getDirectives()
 				.get(Namespace.REQUIREMENT_RESOLUTION_DIRECTIVE));
-			if (optional && !optionalRoots.contains(requirement.getResource())) {
-				List<Capability> value = new ArrayList<>(firstStageResult);
-				Collections.sort(value, capabilityComparator);
-				return value;
-			} else {
-				List<Capability> secondStageList = findProvidersFromRepositories(requirement, firstStageResult);
-
+			if (!optional || optionalRoots.contains(requirement.getResource())) {
+				List<Capability> secondStageResult = findProvidersFromRepositories(requirement, firstStageResult);
 				// Concatenate both stages, eliminating duplicates between the
 				// two
-				firstStageResult.addAll(secondStageList);
-				return new ArrayList<>(firstStageResult);
+				firstStageResult.addAll(secondStageResult);
 			}
+			List<Capability> result = firstStageResult.stream()
+				.sorted(capabilityComparator)
+				.collect(toList());
+			return result;
 		});
-		List<Capability> result = new ArrayList<>(cached);
-		log.log(LogService.LOG_DEBUG, "for " + requirement + " found " + result);
-		return result;
+		List<Capability> capabilities = new ArrayList<>(cached);
+		log.log(LogService.LOG_DEBUG, "for " + requirement + " found " + capabilities);
+		return capabilities;
 	}
 
 	protected void processMandatoryResource(Requirement requirement, LinkedHashSet<Capability> firstStageResult,
@@ -250,35 +250,30 @@ public abstract class AbstractResolveContext extends ResolveContext {
 
 	protected ArrayList<Capability> findProvidersFromRepositories(Requirement requirement,
 		LinkedHashSet<Capability> existingWiredCapabilities) {
-		// Second stage results: repository contents; may be reordered.
-		ArrayList<Capability> secondStageResult = new ArrayList<>();
+		// Second stage results: repository contents.
+		Set<Capability> set = new LinkedHashSet<>();
 
 		// Iterate over the repos
 		int order = 0;
-		ArrayList<Capability> repoCapabilities = new ArrayList<>();
 		for (Repository repo : repositories) {
-			repoCapabilities.clear();
-			Collection<Capability> capabilities = findProviders(repo, requirement);
-			if (capabilities != null && !capabilities.isEmpty()) {
-				repoCapabilities.ensureCapacity(capabilities.size());
-				for (Capability capability : capabilities) {
-					if (isPermitted(capability.getResource()) && ResourceUtils.isEffective(requirement, capability)) {
-						repoCapabilities.add(capability);
+			for (Capability capability : findProviders(repo, requirement)) {
+				if (isPermitted(capability.getResource()) && ResourceUtils.isEffective(requirement, capability)) {
+					if (set.add(capability)) {
 						setResourcePriority(order, capability.getResource());
 					}
 				}
-				secondStageResult.addAll(repoCapabilities);
 			}
 			order++;
 		}
-		Collections.sort(secondStageResult, capabilityComparator);
 
 		// Convert second-stage results to a list and post-process
-		ArrayList<Capability> secondStageList = new ArrayList<>(secondStageResult);
+		ArrayList<Capability> capabilities = set.stream()
+			.sorted(capabilityComparator)
+			.collect(toCollection(ArrayList::new));
 
 		// Post-processing second stage results
-		postProcessProviders(requirement, existingWiredCapabilities, secondStageList);
-		return secondStageList;
+		postProcessProviders(requirement, existingWiredCapabilities, capabilities);
+		return capabilities;
 	}
 
 	/**
@@ -292,12 +287,7 @@ public abstract class AbstractResolveContext extends ResolveContext {
 	 */
 	protected Collection<Capability> findProviders(Repository repo, Requirement requirement) {
 		Map<Requirement, Collection<Capability>> map = repo.findProviders(Collections.singleton(requirement));
-
-		if (map.isEmpty())
-			return Collections.emptySet();
-
 		Collection<Capability> caps = map.get(requirement);
-
 		caps.removeIf(capability -> blacklistedResources.contains(capability.getResource()));
 		return caps;
 	}

--- a/biz.aQute.resolve/test/biz/aQute/resolve/BndrunResolveContextTest.java
+++ b/biz.aQute.resolve/test/biz/aQute/resolve/BndrunResolveContextTest.java
@@ -591,15 +591,15 @@ public class BndrunResolveContextTest extends TestCase {
 		List<Capability> providers = context.findProviders(requirement);
 
 		assertEquals(3, providers.size());
-		assertEquals(IO.getFile("testdata/repo4/x.1.jar")
+		assertEquals(IO.getFile("testdata/repo4/x.3.jar")
 			.toURI(),
 			findContentURI(providers.get(0)
 				.getResource()));
-		assertEquals(IO.getFile("testdata/repo4/x.3.jar")
+		assertEquals(IO.getFile("testdata/repo4/x.2.jar")
 			.toURI(),
 			findContentURI(providers.get(1)
 				.getResource()));
-		assertEquals(IO.getFile("testdata/repo4/x.2.jar")
+		assertEquals(IO.getFile("testdata/repo4/x.1.jar")
 			.toURI(),
 			findContentURI(providers.get(2)
 				.getResource()));

--- a/biz.aQute.resolve/test/biz/aQute/resolve/ResolveTest.java
+++ b/biz.aQute.resolve/test/biz/aQute/resolve/ResolveTest.java
@@ -622,8 +622,8 @@ public class ResolveTest extends TestCase {
 			System.out.println(runbundles);
 			assertThat(bndrun.check()).isTrue();
 			Parameters p = new Parameters(runbundles);
-			assertThat(p.keySet()).hasSize(5)
-				.contains("org.apache.felix.scr", "test.log", "org.apache.felix.log.extension",
+			assertThat(p.keySet()).contains("org.apache.felix.scr", "test.log",
+				"org.apache.felix.log.extension",
 					"org.apache.felix.gogo.command", "org.apache.felix.gogo.runtime");
 		}
 	}

--- a/bndtools.core.test/test.shared.bndrun
+++ b/bndtools.core.test/test.shared.bndrun
@@ -322,7 +322,7 @@
 	org.w3c.dom.events;version='[3.0.0,3.0.1)',\
 	org.w3c.dom.smil;version='[1.0.1,1.0.2)',\
 	org.w3c.dom.svg;version='[1.1.0,1.1.1)',\
-	assertj-core;version='[3.18.0,3.18.1)',\
+	assertj-core;version='[3.18.1,3.18.2)',\
 	org.eclipse.core.net;version='[1.3.400,1.3.401)',\
 	org.eclipse.ui.ide.application;version='[1.3.100,1.3.101)',\
 	org.eclipse.urischeme;version='[1.0.100,1.0.101)'

